volume/overlay: always use parent in cow

cirocosta opened the pull request:
Hey,

As captured in concourse/concourse#5033, once we upgraded the version of the
Kernel we use in `ci.concourse-ci.org`, we started getting errors when
performing copy-on-writes of other copy-on-write volumes in overlay.

That's because the way that we deal with creating copy-on-write volumes of
other copy-on-write vols hit the following condition:

> Using an upper layer path and/or a workdir path that are already used by
> another overlay mount is not allowed and may fail with EBUSY.

We can see that by capturing the `mount(2)` syscall data arg:


	COW (of `handle`):
		lowerdir=handle,upperdir=cow-handle,workdir=work/cow-handle
			   |
			   '-> regular bind mount
			   

	COW-COW
		lowerdir=cow-handle:handle,upperdir=cow-cow-handle,workdir=work/cow-cow-handle
			   |
			   '-> oh oh, this is cow's upper dir
			   	--> not good

However, if we switch the logic of searching for the ancestors, and just take
the `parent` mountpoint (which would point to `cow-handle/volume`), we'll be
using a non-upper as `lower` (which is valid), while still getting visibility of
all of the ancestors anyway (as in a chain of `cow`'s, the latest `parent` will
already have performed the whole union of the chain).

relates to https://github.com/concourse/concourse/issues/4964

---

sometime ago I wrote about overlay: https://ops.tips/notes/practical-look-into-overlayfs/
updated the article with this scenario 😁

Commit:
volume/overlay: always use parent in cow

instead of figuring out the whole ancestry chain when performing a
copy-on-write in `overlay`, go with `parent` directly.

this has the effect of always pointing to directories that are
mountpoints, rather than pieces that might constitute `upperdir` of
other volumes.

Signed-off-by: Ciro S. Costa <cscosta@pivotal.io>

vito commented on volume/driver/overlay_linux.go:
```diff
@@ -68,15 +68,10 @@ func (driver *OverlayDriver) DestroyVolume(path string) error {
 }
 
 func (driver *OverlayDriver) CreateCopyOnWriteLayer(path string, parent string) error {
-	ancestry, err := driver.ancestry(parent)
```
Can we kill the `ancestry` method too? I don't see it used anymore (though GitHub search could be failing me).

cirocosta replied:
ooh it's used in the "remount"ing (`RecoverMountTable`): 

https://github.com/concourse/baggageclaim/blob/dc362b1f055e4836dc2afb16639383c61cee637a/volume/driver/overlay_linux.go#L139-L142 

I didn't really look at it yet though

cirocosta replied:
(going to try it out here)

cirocosta replied:
yeahh ... we'll have to modify that too - and it turns out that we can't just go with the `parent` directly: the `mount(2)` syscalls will work, but the data *might* not show up, depending on the order of the mounts.

if it just turns out that in the iteration we get the `cow_cow` before `cow`, we'll be able to mount `cow_cow` just fine, but by mounting `cow` right after, it's as if you had "stacked", thus, `cow_cow` won't see the contents from `cow`.

it seems to me that we'd need to make sure that we mount parents first, and then cow_cow later (which makes sense to me)

reproducible: https://gist.github.com/cirocosta/c2cf9a7b818dd8e88fb67209bb9debd1

vito reviewed (dismissed):
Interesting, looks like I/we misunderstood the capabilities of `overlayfs`. I remember running into trouble with nesting overlay filesystems, but I think the only real problem is that you can't have the *upper* directory live on an overlay mount. The `overlayfs` docs explicitly state that using one as a lower mount is fine:

> The lower filesystem can be any filesystem supported by Linux and does
> not need to be writable.  **The lower filesystem can even be another
> overlayfs.**  The upper filesystem will normally be writable and if it
> is it must support the creation of trusted.* extended attributes, and
> must provide valid d_type in readdir responses, so NFS is not suitable.

(emphasis mine)

So, 👍 from me, though we can probably remove the entire `ancestry` method, so just requested that tiny change.

**edit:** actually looks like we do use `ancestry` here: https://github.com/concourse/baggageclaim/blob/fca485963e6072860050a9ea8c10ef6a582db1c5/volume/driver/overlay_linux.go#L139 - but this code should probably be changed!

Commit:
ovl: re-implement RecoverMountTable w/ ancestry graph traversal

here, RecoverMountTable getss refactored to:

1. have unit tests that verify the behavior
2. perform cow of cow volumes leveraging the parent's mountpoint, rather
   than mounting with the whole mount ancestry as lowerdirs.

ps.: no changes to integration tests were performed.

Signed-off-by: Ciro S. Costa <cscosta@pivotal.io>

xtreme-sameer-vohra commented on volume/driver/overlay_linux.go:
```diff
@@ -196,28 +144,149 @@ func (driver *OverlayDriver) workDir(datapath string) string {
+// It takes care of finding out the dependencies between volumes (represented by
+// symbolic links at `live/<vol>/parent`), and making sure that:
+// - dependencies are mounted first
+// - mounpoints are not mounted more than once
```
nit: mountpoints

xtreme-sameer-vohra commented on volume/driver/overlay_linux.go:
```diff
@@ -196,28 +144,149 @@ func (driver *OverlayDriver) workDir(datapath string) string {
+}
+
+// Ancestry traverses LiveVolume linked list producing a list of LiveVolumes
+// from "no dependencies" to "with dependencies".
```
nit: from "no dependencies" aka oldest to "with dependencies" aka youngest

cirocosta replied:
wdyt?

```diff
- // from "no dependencies" to "with dependencies". 
+ // from "no dependencies" (oldest) to "with dependencies" (newest). 
```

xtreme-sameer-vohra commented on volume/driver/overlay_linux.go:
```diff
@@ -196,28 +144,149 @@ func (driver *OverlayDriver) workDir(datapath string) string {
+// it figures out an acceptable mounting order is `vol1`, `vol3`, `vol2`,
+// `vol4`, issuing precisely only 4 `mount`s.
+//
+func RecoverMountTable(root string, mounter Mounter) error {
```
Its great this isn't part of the interface and is standalone 👍

xtreme-sameer-vohra commented on volume/driver/overlay_linux.go:
```diff
@@ -5,14 +5,25 @@ import (
+	Parent *LiveVolume
+}
+
+type Mounter interface {
```
Assuming this was pulled out to simplify testing

xtreme-sameer-vohra commented on volume/driver/overlay_linux.go:
```diff
@@ -196,28 +144,149 @@ func (driver *OverlayDriver) workDir(datapath string) string {
-func (driver *OverlayDriver) getGUID(datapath string) string {
-	return filepath.Base(filepath.Dir(datapath))
+func readlink(path string) (target string, err error) {
+	finfo, err := os.Lstat(path)
```
nice to move this inside readlink

xtreme-sameer-vohra commented on volume/driver/overlay_linux_test.go:
```diff
@@ -0,0 +1,330 @@
+				})
+			})
+
+			Context("having a file in the middle of vol dirs", func() {
```
nit: having a random dir in the root dir rather than file, as file would be ignored 🤔

xtreme-sameer-vohra commented on volume/driver/overlay_linux_test.go:
```diff
@@ -0,0 +1,330 @@
+					Expect(err).ToNot(HaveOccurred())
+				})
+
+				It("fails", func() {
```
Seems a bit brittle to error on this.
WDYT ?

cirocosta replied:
My thinking was on the lines of "if there's a file here, something went reeeaally wrong, possibly in the config" - perhaps the user misconfigured the location on where volumes live.

By failing here (in a method that's used only during startup time), the user can get that feedback that something is on the way, rather than later through the silence 🤔

xtreme-sameer-vohra commented on volume/driver/overlay_linux_test.go:
```diff
@@ -0,0 +1,330 @@
+			res, err = NewLiveVolume(root, vol)
+		})
+
+		Context("with inexistent root", func() {
```
nit: non existent

cirocosta replied:
https://www.dictionary.com/browse/inexistent 🤔 maybe it's a more 'latin'-based form? (in Portuguese, "inexistent" is "inexistente" haha super close)

xtreme-sameer-vohra replied:
whoops

xtreme-sameer-vohra commented on volume/driver/overlay_linux.go:
```diff
@@ -196,28 +144,149 @@ func (driver *OverlayDriver) workDir(datapath string) string {
-		}
+			datapath := filepath.Join(ancestor.Path, "volume")
+
+			if idx == 0 { // volume with no parents (regular vol)
```
we could check the ancestor.Parent == nil to decide

xtreme-sameer-vohra replied:
[@cirocosta w/ @xtreme-sameer-vohra] : we could actually do some refactoring here to do something like this:

```go


type MountTable struct {
  Mounter Mounter

  storage map[string]interface{}
}

func (m MountTable) isAlreadyMounted(vol LiveVolume) bool {
  _, found := m.storage[vol.Path]
  return found
}


func (m MountTable) markAsMounted(vol LiveVolume) bool {
  _, found := m.storage[vol.Path]
  return found
}

func (m MountTable) Mount(vol LiveVolume) {
  if m.isAlreadyMounted(vol) {
    return
  }

  if vol.Parent == nil {
    m.Mounter.BindMount()
    m.MarkAsMounted(vol)
    return
  }

  m.Mounter.OverlayMount(vol)
  m.MarkAsMounted(vol)
}
```

cirocosta replied:
it ends up looking like this:

```go
func RecoverMountTable(root string, mounter Mounter) error {
        dirs, err := ioutil.ReadDir(root)
        if err != nil {
                if os.IsNotExist(err) {
                        return nil
                }

                err = fmt.Errorf("readdir %s: %w", root, err)
                return err
        }

        table := NewMountTable(mounter)

        for _, dir := range dirs {
                vol, err := NewLiveVolume(root, dir.Name())
                if err != nil {
                        return err
                }

                ancestry := Ancestry(*vol)
                for _, ancestor := range ancestry {
                        err = table.Mount(ancestor)
                        if err != nil {
                                return fmt.Errorf("mounting vol %+v: %w", ancestor, err)
                        }
                }
        }

        return nil
}
```

kcmannem replied:
clean!

xtreme-sameer-vohra reviewed (changes requested):
lgtm !!

Commit:
volume/overlay: address nits in comments

Signed-off-by: Ciro S. Costa <cscosta@pivotal.io>

Commit:
volume/overlay: reefactor recovermounttable

rather than having all of the decisions  around how (and whether to)
mounting inline, now we hav `MountTable` who's responsible for just
that, making the overall idea of what "recovering a mounting table" more
straightforward: go over the liv vols, and mount them.

Signed-off-by: Ciro S. Costa <cscosta@pivotal.io>

vito commented:
@xtreme-sameer-vohra Did you mean to approve? :stuck_out_tongue:

izabelacg commented:
yes, but @cirocosta and I had ended up discussing another refactor which is now in !!

izabelacg commented on volume/driver/overlay_linux_test.go:
```diff
@@ -14,7 +14,7 @@ import (
 )
 
-var _ = Describe("Overlay", func() {
+var _ = FDescribe("Overlay", func() {
```
Hey, it snuck in 👀

izabelacg committed:
Fixed typo and removed unused library

Signed-off-by: Sameer Vohra <svohra@pivotal.io>

xtreme-sameer-vohra reviewed (approved):
@cirocosta thx for the refactor
@pivotal-izabela-gomes thx for fixing the typo

Commit:
volume/overlay: add driver import to tests

previously, we had `driver` coming as a dot import. Once that was
removed, the tests didn't compile anymore.

Here I bring driver back, but as a regular import this time.

Signed-off-by: Ciro S. Costa <cscosta@pivotal.io>
